base/linalg/dense.jl

```diff
@@ -309,17 +309,64 @@ kron(a::AbstractVector, b::AbstractVector)=vec(kron(reshape(a,length(a),1),resha
 kron(a::AbstractMatrix, b::AbstractVector)=kron(a,reshape(b,length(b),1))
 kron(a::AbstractVector, b::AbstractMatrix)=kron(reshape(a,length(a),1),b)
 
-^(A::AbstractMatrix, p::Integer) = p < 0 ? inv(A^-p) : Base.power_by_squaring(A,p)
-
-function ^(A::AbstractMatrix, p::Number)
+# Matrix power
+^{T}(A::AbstractMatrix{T}, p::Integer) = p < 0 ? Base.power_by_squaring(inv(A),-Integer(p)) : Base.power_by_squaring(A,Integer(p))
+function ^{T}(A::AbstractMatrix{T}, p::Real)
+    n = checksquare(A)
+    # Quick return if A is diagonal
+    if isdiag(A)
+        for i in 1:n
+            A[i, i] = A[i, i] ^ p
+        end
+        return A
+    end
+    # For integer powers, use repeated squaring
     if isinteger(p)
-        return A^Integer(real(p))
+        TT = Base.promote_op(^, eltype(A), typeof(p))
+        return (TT == eltype(A) ? A : copy!(similar(A, TT), A))^Integer(p)
+    end
+
+    # If possible, use diagonalization
+    if issymmetric(A) && T <: Real
+        return full(Symmetric(A)^p)
+    end
+    if ishermitian(A)
+        return full(Hermitian(A)^p)
+    end
+
+    # Otherwise, use Schur decomposition
+    if istriu(A)
+        #Integer part
+        retmat = full(A) ^ floor(p)
+        #Real part
+        retmat = retmat * full(powm(UpperTriangular(float.(A)), real(p - floor(p))))
+        d = diag(A)
+    else
+        S,Q,d = schur(complex(A))
+        #Integer part
+        R = full(S) ^ floor(p)
+        #Real part
+        R = R * full(powm(UpperTriangular(float.(S)), real(p - floor(p))))
+        retmat = Q * R * Q'
+    end
+
+    # Check whether the matrix has nonpositive real eigs
+    np_real_eigs = false
+    for i = 1:n
+        if imag(d[i]) < eps() && real(d[i]) <= 0
```

Comment:
This is not a good way to do the test because it depends on overall scaling of the matrix (if you multiply the whole matrix by `1e±100` the result will change).

Also, shouldn't you be checking `abs(imag)`, not `imag`?  And shouldn't the warning be for `real < 0`, not real ≤ 0`?

I'm not sure what a robust test would be here.  Maybe `abs(imag(d[i])) ≤ ɛ`, where `ɛ = eps(maxabs(d))`?